plugin/core/views.py

```diff
@@ -425,7 +425,15 @@ def text_document_identifier(view_or_uri: Union[DocumentUri, sublime.View]) -> T
     return {"uri": uri}
 
 
+def get_first_selection_region(selelection: sublime.Selection) -> Optional[sublime.Region]:
```

Comment:
Is this function really necessary? Now you need to put `view.sel()` into the argument at all the places where it is called.
And I think you can rewrite the relevant change from this PR in a simpler way without the need to use this function
```python
selection = view.sel()
if selection:
    current_first_region = selection[0]
    ...
```

Reply:
And if we want to avoid calling it multiple times then i needed to change it like that. Though I'm not convinced it was necessary to options that in the first place

Reply:
Note that `len` does a call to a native API so it would be in practice less efficient. Does it matter? Probably not.

```py
    def __len__(self):
        return sublime_api.view_selection_size(self.view_id)

    def __getitem__(self, index):
        r = sublime_api.view_selection_get(self.view_id, index)
        if r.a == -1:
            raise IndexError()
        return r
```

Reply:
Well, I figured I should test instead of guessing.

```py
    def test_len():
        return sel[0] if len(sel) else None

    repeat = 100000
    times = timeit.repeat(test_len, repeat=3, number=repeat)
    return 'Best of {}: {} usec.'.format(repeat, min(times))

	# 0.26699970833396947 usec.
```

```py
    def test_trycatch():
        try:
            return sel[0]
        except Exception:
            return None

    repeat = 100000
    times = timeit.repeat(test_trycatch, repeat=3, number=repeat)
    return 'Best of {}: {} usec.'.format(repeat, min(times))

	# 0.4004749166670081 usec.
```

So no, `len` variant is faster.

Reply:
I rewrote `_update_stored_selection_async` in a different way.

Reply:
Weird. I get different benchmark results.

```py
import sublime
import sublime_plugin
import timeit


class ExampleCommand(sublime_plugin.TextCommand):
    def run(self, edit: sublime.Edit) -> None:
        sel = self.view.sel()
        repeat = 100000

        def test_len_1():
            return sel[0] if len(sel) else None

        def test_len_2():
            try:
                return sel[0]
            except Exception:
                return None

        def test_len_3():
            return next(iter(sel), None)

        times = timeit.repeat(test_len_1, repeat=3, number=repeat)
        print("(test_len_1) Best of {}: {} usec.".format(repeat, min(times)))
        times = timeit.repeat(test_len_2, repeat=3, number=repeat)
        print("(test_len_2) Best of {}: {} usec.".format(repeat, min(times)))
        times = timeit.repeat(test_len_3, repeat=3, number=repeat)
        print("(test_len_3) Best of {}: {} usec.".format(repeat, min(times)))
```

results in 

```
(test_len_1) Best of 100000: 0.39621629999999186 usec.
(test_len_2) Best of 100000: 0.2581017000000543 usec.
(test_len_3) Best of 100000: 0.2883709000000181 usec.
```

Reply:
I can see the same here when using actual `view.sel` rather than manually created `sublime.Selection(1).

```
* view.sel()
(test_len_1) Best of 100000: 0.26706908333332535 usec.
(test_len_2) Best of 100000: 0.4003557916666409 usec.
(test_len_3) Best of 100000: 0.4005454583333119 usec.

* sublime.Selection(1)
(test_len_1) Best of 100000: 0.5351135833333274 usec.
(test_len_2) Best of 100000: 0.26820062500001995 usec.
(test_len_3) Best of 100000: 0.2768708750000428 usec.
```

Reply:
Which seems to boil down to `__len__` implementation having to do an API call and that appears to have more overhead when view ID is not valid (does not exist).

Reply:
Any more comments on this?